### ISSUE TITLE
BackgroundSubtractorKNN does not respect a zero learning rate (same with GSOC and LSBP)

### ISSUE THREAD

**Issue by ArdooTala** [opencv/opencv#24226]
### System Information

OpenCV python version: 4.8.0.76 opencv-contrib-python
Operating System / Platform: Ubuntu 22.04
Python version: 3.10.12

### Detailed description

The `BackgroundSubtractorKNN()` does **not** stop updating the background model regardless of the zero `learningRate` factor when calling the `BackgroundSubtractorKNN.apply(<image>, None, learningRate=0)`.
It works fine when using the `BackgroundSubtractorMOG2`, but the learningRate param in `BackgroundSubtractorKNN` does not respect the learning rate. The background model keeps updating regardless of the value.

The same issue exists with `BackgroundSubtractorLSBP` and `BackgroundSubtractorGSOC` from the *cv2.bgsegm* package. Although `BackgroundSubtractorMOG` and `BackgroundSubtractorMGM` work fine.

### Steps to reproduce

```python
import cv2


back_sub = cv2.createBackgroundSubtractorKNN()
# back_sub = cv2.bgsegm.createBackgroundSubtractorLSBP()    # Has the same problem
# back_sub = cv2.bgsegm.createBackgroundSubtractorGSOC()    # Has the same problem

cap = cv2.VideoCapture(0)
if not cap.isOpened():
    raise (Exception("Could not open capture"))

learning_rate = 0.1
while True:
    ret, frame = cap.read()
    if not ret:
        break

    fg_mask = back_sub.apply(frame, None, learning_rate)
    print(learning_rate)

    cv2.imshow("frame", frame)
    cv2.imshow("foreground", fg_mask)
    cv2.imshow("background", back_sub.getBackgroundImage())

    k = cv2.waitKey(1)
    if k == ord("q"):
        break
    # Press "d" to switch learning_rate between 0.1 and 0.0
    elif k == ord("d"):
        if not learning_rate:
            learning_rate = 0.1
        else:
            learning_rate = 0.0

cap.release()
cv2.destroyAllWindows()
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by LaurentBerger** [comment]
It's an opencv_contrib issue

**Comment by Suyashd999** [comment]
Digging in the code (cv2-> `__init__.pyi`), the BackgroundSubtractorMOG2 class has the following functions defined:

```
@typing.overload
def apply(self, image: cv2.typing.MatLike, fgmask: cv2.typing.MatLike | None = ..., learningRate: float = ...) -> cv2.typing.MatLike: ...
@typing.overload
def apply(self, image: UMat, fgmask: UMat | None = ..., learningRate: float = ...) -> UMat: ...
```

These function definitions are not present in the `BackgroundSubtractorKNN` class. This might be the reason the learning rate might not be getting applied.  I am just a newbie so was not exactly sure how to fix it.

**Comment by JustASquid** [comment]
Running into this one too. Very problematic for my application.

**Comment by LGJLI** [comment]
Could this be the culprit ? 
https://github.com/opencv/opencv/blob/7cf075c3924b57420484a59829f47af72274c8dd/modules/video/src/bgfg_KNN.cpp#L760-L770

What happens with Kshort, Kmid and Klong if learningRate == 0 ?

**Comment by Carlos121084** [comment]
Same issue for me
MOG2 works fine, but KNN keeps learning with the learningrate set to 0
Only "workaround" I found is to set the learningrate to 0.0000000001
